### PR TITLE
Ndcube 2.0 fixes

### LINKED ISSUE #12: Use ndcube >= 2.0
We chatted about this on the PyAstro telecon - but did we decide to use NDCube master or 1.3.2?

### PR DESCRIPTION
**Pull request by wtbarnes**
### Description

Fixes #12 

A few more things to do still 

- [x] Fix temperature axis physical type
- [ ] ~~Add base model tests~~
- [ ] ~~CI to make sure these tests actually pass~~

### CODE REVIEW

File: sunkit_dem/base_model.py [hunk 1]
```diff
@@ -110,24 +112,22 @@ def kernel(self):
     def kernel(self, kernel):
         if len(kernel) != self.data.cube_like_dimensions[0].value:
             raise ValueError('Number of kernels must be equal to length of wavelength dimension.')
-        if not all([kernel[k].shape == self.temperature_bin_centers.shape for k in kernel]):
+        if not all([k.shape == self.temperature_bin_centers.shape for k in kernel]):
```
Comment (PaulJWright) [inline sunkit_dem/base_model.py:115]: I was working off this branch using NDCube 2.0.0, and had to revert this change for things to work?

I was getting the error:

```
AttributeError: 'int' object has no attribute 'shape'
```

`kernel` in my case is the AIA response functions [94, 131, 171, 193, 211, 335]

```
print(type(kernel), len(kernel))
<class 'dict'> 6

[k for k in response]
[94, 131, 171, 193, 211, 335]

[kernel[k].shape == temperature_bin_centers.shape for k in kernel]
[True, True, True, True, True, True]
```

File: sunkit_dem/base_model.py [hunk 2]
```diff
@@ -110,24 +112,22 @@ def kernel(self):
...
 
     @property
     def data_matrix(self):
         return u.Quantity(np.stack([n.data.squeeze() for n in self.data]), self.data[0].unit)
 
     @property
     def kernel_matrix(self):
-        return u.Quantity(np.stack([self.kernel[w].value for w in self.wavelength.value]),
-                          self.kernel[self.wavelength.value[0]].unit)
+        return u.Quantity(np.stack([k.value for k in self.kernel]), self.kernel[0].unit)
```
Comment (PaulJWright) [inline sunkit_dem/base_model.py:130]: As with my above comment, I also had to change this back to:

```

    @property
    def kernel_matrix(self):
        first_key = list(self.kernel)[0]
        return u.Quantity(
            np.stack([self.kernel[k].value for k in self.kernel]),
            self.kernel[first_key].unit,
        )
```

**Comment by PaulJWright** [comment]
With respect to the first additional change: "Fix temperature axis physical type", what needs to be done here?

**Comment by wtbarnes** [comment]
This PR now does a few more things:

- Data input is now an `NDCollection`, not a sequence.
- Temperature axis can now have arbitrary spacing as the WCS is not limited to FITS WCS
- I've added a rough version of the Hannah and Kontar model taken from here: https://github.com/ianan/demreg/tree/master/python. This also includes a change to use Dask for parallelization.

The last point change should probably go into it's own PR or live in an entirely separate package.

File: sunkit_dem/base_model.py (hunk 2 above)
Comment (wtbarnes) [reply sunkit_dem/base_model.py]: I think this should now be fixed as I'm constructing the kernel matrix a bit differently and being careful to preserve the correct ordering between the data matrix and the kernel matrix.

**Comment by PaulJWright** [comment]
> This PR now does a few more things:
> 
> * Data input is now an `NDCollection`, not a sequence.
> * Temperature axis can now have arbitrary spacing as the WCS is not limited to FITS WCS
> * I've added a rough version of the Hannah and Kontar model taken from here: https://github.com/ianan/demreg/tree/master/python. This also includes a change to use Dask for parallelization.
> 
> The last point change should probably go into it's own PR or live in an entirely separate package.

I'll take a look! 

For the last point, yes, I agree. But if we have to include it here for now to make the package useful, then I'm fine with that?

**Comment by PaulJWright** [comment]
The changes seem to make sense. Let me know when this is out of draft and I can properly go through it! :)

**Comment by PaulJWright** [comment]
Can we merge this and create issues for the following items?

-  Add base model tests
-  CI to make sure these tests actually pass

What needs to be done to "Fix temperature axis physical type"?

**Comment by wtbarnes** [comment]
> What needs to be done to "Fix temperature axis physical type"? I am happy to have a go if I know where to start

Ha I realized that I repeatedly did not reply to this. I think this is not an issue now that the temperature axis is not limited to FITS WCS

**Comment by wtbarnes** [comment]
I'm going to go ahead and merge this as is as I think this largely addresses the initial problem of compatibility with ndcube 2.0. I've created issues for the remaining TODOs
